src/wasm-traversal.h

```diff
@@ -0,0 +1,420 @@
+
+  // children-first
+  void walk(Expression*& curr) override {
+    if (!curr) return;
+
+    // special-case Block, because Block nesting (in their first element) can be incredibly deep
+    if (curr->is<Block>()) {
+      auto* block = curr->dyn_cast<Block>();
+      std::vector<Block*> stack;
+      stack.push_back(block);
+      while (block->list.size() > 0 && block->list[0]->is<Block>()) {
+        block = block->list[0]->cast<Block>();
+        stack.push_back(block);
+      }
+      // walk all the children
+      for (int i = int(stack.size()) - 1; i >= 0; i--) {
+        auto* block = stack[i];
+        auto& children = block->list;
+        for (size_t j = 0; j < children.size(); j++) {
+          if (i < int(stack.size()) - 1 && j == 0) {
+            // this is one of the stacked blocks, no need to walk its children, we are doing that ourselves
+            this->visit(children[0]);
+            if (replace) {
+              children[0] = replace;
+              replace = nullptr;
+            }
+          } else {
+            this->walk(children[j]);
+          }
+        }
+      }
+      // we walked all the children, and can rejoin later below to visit this node itself
+    } else {
+      // generic child-walking
+      ChildWalker<WasmWalker<SubType, ReturnType>>(*this).visit(curr);
+    }
+
+    this->visit(curr);
+
+    if (replace) {
+      curr = replace;
+      replace = nullptr;
+    }
+  }
+
+  void startWalk(Function *func) override {
+    walk(func->body);
+  }
+
+  void startWalk(Module *module) override {
+    // Dispatch statically through the SubType.
+    SubType* self = static_cast<SubType*>(this);
+    for (auto curr : module->functionTypes) {
+      self->visitFunctionType(curr);
+      assert(!replace);
+    }
+    for (auto curr : module->imports) {
+      self->visitImport(curr);
+      assert(!replace);
+    }
+    for (auto curr : module->exports) {
+      self->visitExport(curr);
+      assert(!replace);
+    }
+    for (auto curr : module->functions) {
+      startWalk(curr);
+      self->visitFunction(curr);
+      assert(!replace);
+    }
+    self->visitTable(&module->table);
+    assert(!replace);
+    self->visitMemory(&module->memory);
+    assert(!replace);
+    self->visitModule(module);
+    assert(!replace);
+  }
+};
+
+// Traversal in the order of execution. This is quick and simple, but
+// does not provide the same comprehensive information that a full
+// conversion to basic blocks would. What it does give is a quick
+// way to view straightline execution traces, i.e., that have no
+// branching. This can let optimizations get most of what they
+// want without the cost of creating another AST.
+//
+// When execution is no longer linear, this notifies via a call
+// to noteNonLinear().
+
+struct FastExecutionWalker : public WasmWalkerBase<FastExecutionWalker> {
+  FastExecutionWalker() {}
+
+  void noteNonLinear() { abort(); } // must be overridden
```

Comment:
But this class isn't a template. If you're going to derive from it, and you want these base class methods to call `noteNonLinear` in the derived class, then it has to be virtual (otherwise they will call the abort implementation). If you want CRTP all the way down then this class has to be a template too.

Backing up, what would the derived class be expected to implement?